libpromises/eval_context.c

```diff
@@ -1950,24 +1954,65 @@ StringSet *EvalContextStackPromisees(const EvalContext *ctx)
     return promisees;
 }
 
+/**
+ * We cannot have double-scoped variables (e.g. "this.config.var1"), so if we
+ * want to put a scoped variable into a special scope, we need to mangle the
+ * name like this:
+ *   "config.var1" -> "config___var1"
+ */
+static inline char *MangleScopedVarNameIntoSpecialScopeName(const char *scope, const char *var_name)
+{
+    const size_t var_name_len = strlen(var_name);
+
+    /* Replace '.' with NESTED_SCOPE_SEP */
+    char *new_var_name = xmalloc(var_name_len + sizeof(NESTED_SCOPE_SEP));
+    memcpy(new_var_name, var_name, var_name_len + 1 /* including '\0' */);
+
+    /* Make sure we only replace the "scope." string, not all dots. */
+    char *scope_with_dot = StringConcatenate(2, scope, ".");
+    char *scope_with_underscores = StringConcatenate(2, scope, NESTED_SCOPE_SEP);
+
+    NDEBUG_UNUSED ssize_t ret = StringReplace(new_var_name, var_name_len + sizeof(NESTED_SCOPE_SEP),
+                                              scope_with_dot, scope_with_underscores);
+    assert(ret == (var_name_len + sizeof(NESTED_SCOPE_SEP) - 2));
+
+    free(scope_with_dot);
+    free(scope_with_underscores);
+
+    return new_var_name;
+}
+
 /*
  * Copies value, so you need to free your own copy afterwards.
  */
 bool EvalContextVariablePutSpecial(EvalContext *ctx, SpecialScope scope, const char *lval, const void *value, DataType type, const char *tags)
 {
+    char *new_lval = NULL;
+    if (strchr(lval, '.') != NULL)
+    {
+        VarRef *ref = VarRefParse(lval);
+        if (ref->scope != NULL)
+        {
+            new_lval = MangleScopedVarNameIntoSpecialScopeName(ref->scope, lval);
+        }
+        VarRefDestroy(ref);
+    }
     if (strchr(lval, '['))
     {
         // dealing with (legacy) array reference in lval, must parse
-        VarRef *ref = VarRefParseFromScope(lval, SpecialScopeToString(scope));
+        VarRef *ref = VarRefParseFromScope(new_lval ? new_lval : lval, SpecialScopeToString(scope));
```

Comment:
CONTRIBUTING.md says to use explicit comparisons. I.e. `new_lval != NULL` instead of ~~`new_lval`~~. Although I'm not sure if this applies here.

Reply:
I was actually thinking about this. But I think it's just better this way in the tertiary operator because the explicit comparison adds a lot of noise.

Reply:
Sure, I agree